rust/tool/src/systemd.rs

```diff
@@ -0,0 +1,84 @@
+use std::ffi::CStr;
+use std::fs;
+use std::path::Path;
+use std::str::FromStr;
+
+use anyhow::{Context, Result};
+
+use crate::os_release::OsRelease;
+use crate::pe;
+
+/// A systemd version.
+///
+/// The version is parsed into a u32 tuple because systemd does not follow strict semver
+/// conventions. A major version without a minor version, e.g. "252" is represented as `(252, 0)`.
+#[derive(PartialEq, PartialOrd, Debug)]
+pub struct SystemdVersion(u32, u32);
+
+impl SystemdVersion {
+    /// Read the systemd version from the `.osrel` section of a systemd-boot binary.
+    pub fn from_systemd_boot_binary(path: &Path) -> Result<Self> {
+        let file_data = fs::read(path).with_context(|| format!("Failed to read file {path:?}"))?;
+        let section_data = pe::read_section_data(&file_data, ".osrel")
+            .with_context(|| format!("PE section '.osrel' is empty: {path:?}"))?;
+
+        // The `.osrel` section in the systemd-boot binary is a NUL terminated string and thus needs
+        // special handling.
+        let section_data_cstr =
+            CStr::from_bytes_with_nul(section_data).context("Failed to parse C string.")?;
+        let section_data_string = section_data_cstr
+            .to_str()
+            .context("Failed to convert C string to Rust string.")?;
+
+        let os_release = OsRelease::from_str(section_data_string)
+            .with_context(|| format!("Failed to parse os-release from {section_data_string}"))?;
+
+        let version_str = os_release
+            .0
+            .get("VERSION")
+            .context("Failed to extract VERSION key from: {os_release:#?}")?;
+
+        Self::from_str(version_str)
+    }
+
+    /// Parse a systemd version from a string.
+    pub fn from_str(s: &str) -> Result<Self> {
+        let mut split_version_string = s.split('.');
```

Comment:
Couldn't this by `s.split(...)..map(... u32::from_str ...).collect::<Result<Vector<u32>>()?;` 

Then you can also check whether you actually get 2 elements and fix as required. If the version number tends to have crap at the end you can also add a `.take(2)` above.